File: solutions/day3.ts
import fs from "node:fs";

const raw = fs.readFileSync("./input.txt", "utf-8");

// Part 1
const regexPart1 = /mul\(([0-9]+),([0-9]+)\)/g;
const sumPart1 = [...raw.matchAll(regexPart1)]
  .map((match) => parseInt(match[1]) * parseInt(match[2]))
  .reduce((acc, curr) => acc + curr, 0);
console.log("Part 1:", sumPart1);

// Part 2
const regexPart2 = /mul\(([0-9]+),([0-9]+)\)|don?\'?t?\(\)/g;
let isDisabled = false;
const sumPart2 = [...raw.matchAll(regexPart2)].reduce<number>((acc, match) => {
  if (match[1] && !isDisabled) {
    return acc + parseInt(match[1]) * parseInt(match[2]);
  } else if (match[0] === "don't()") {
    isDisabled = true;
  } else if (match[0] === "do()") {
    isDisabled = false;
  }
  return acc;
}, 0);
console.log("Part 2:", sumPart2);

File: solutions/python/day3.py
import re

with open("./input.txt", "r") as txt:
  mem: str = txt.read().rstrip()
  txt.close()

# Part 1
resultsPart1 = re.findall(r"mul\(([0-9]+),([0-9]+)\)", mem)
sumPart1 = 0
for i in range(len(resultsPart1)):
  sumPart1 += int(resultsPart1[i][0]) * int(resultsPart1[i][1])
print("Part 1:", sumPart1)

# Part 2
resultsPart2 = re.findall(r"mul\(([0-9]+),([0-9]+)\)|(don?\'?t?\(\))", mem)
sumPart2 = 0
isDisabled = False
for i in range(len(resultsPart2)):
  if re.fullmatch(r"[0-9]+", resultsPart2[i][0]) != None and not isDisabled:
    sumPart2 += int(resultsPart2[i][0]) * int(resultsPart2[i][1])
  elif re.fullmatch(r"don\'t\(\)", resultsPart2[i][2]) != None:
    isDisabled = True
  elif re.fullmatch(r"do\(\)", resultsPart2[i][2]) != None:
    isDisabled = False
print("Part 2:", sumPart2)
